src/quarot/modeling_llama.py

```diff
@@ -0,0 +1,151 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT license.
+
+from typing import Optional, Tuple
+
+import torch
+from transformers import Cache, LlamaConfig
+from transformers.models.llama.modeling_llama import (
+    LlamaFlashAttention2,
+    LlamaForCausalLM,
+    LlamaMLP,
+    apply_rotary_pos_emb,
+)
+from transformers.pytorch_utils import ALL_LAYERNORM_LAYERS
+
+from quarot.nn import DummyActQuantizer, OnlineHadamard, QuarotFP16Linear
+from slicegpt.modules import RMSN
+
+ALL_LAYERNORM_LAYERS.append(RMSN)
+
+
+class QuarotLlamaConfig(LlamaConfig):
+    model_type = "llama_quarot"
+
+
+class QuarotLlamaMLP(LlamaMLP):
+    def __init__(self, config, act_quantization=False, online_had=False, *args, **kwargs):
+        super().__init__(config, *args, **kwargs)
+        self.act_quantization = act_quantization
+        self.online_had = online_had
+        self.input_quantizer = DummyActQuantizer()
+        self.up_proj = QuarotFP16Linear.like(self.up_proj)
+        self.gate_proj = QuarotFP16Linear.like(self.gate_proj)
+        self.down_proj = QuarotFP16Linear.like(self.down_proj)
+        self.online_down_proj_hadamard = OnlineHadamard(self.intermediate_size)
+        self.down_proj_input_quantizer = DummyActQuantizer()
+
+    def forward(self, x):
+        # Quantize inputs to mlp
+        x = self.input_quantizer(x)
+
+        # Calculate activations
+        x = self.act_fn(self.gate_proj(x)) * self.up_proj(x)
+
+        # Apply online hadamard if needed
+        if self.online_had:
+            x = self.online_down_proj_hadamard(x)
+
+        # Quantize inputs to down_proj
+        x = self.down_proj_input_quantizer(x)
+
+        # Return final activations
+        return self.down_proj(x)
+
+
+class QuarotFP16LlamaFlashAttention2(LlamaFlashAttention2):
```

Comment:
Is this file the replacement for monkeypatch.py?

Reply:
Yes. We either monkeypatch or we specify the forward passes ourselves.